plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,402 @@
+            Map.insertWith max fun (length args)
+      _ -> id
+
+-- | The second pass. See Note [CSE].
+annotate :: Term name uni fun ann -> Term name uni fun (Path, ann)
+annotate = flip evalState 0 . flip runReaderT [] . go
+  where
+    -- The integer state is the highest ID assigned so far.
+    -- The reader context is the current path.
+    go :: Term name uni fun ann -> ReaderT Path (State Int) (Term name uni fun (Path, ann))
+    go t = do
+      path <- ask
+      case t of
+        Apply ann fun arg -> Apply (path, ann) <$> go fun <*> go arg
+        Force ann body -> Force (path, ann) <$> go body
+        Constr ann i args -> Constr (path, ann) i <$> traverse go args
+        Constant ann val -> pure $ Constant (path, ann) val
+        Error ann -> pure $ Error (path, ann)
+        Builtin ann fun -> pure $ Builtin (path, ann) fun
+        Var ann name -> pure $ Var (path, ann) name
+        LamAbs ann n body -> do
+          freshId <- (+ 1) <$> lift get
+          lift $ put freshId
+          LamAbs (path, ann) n <$> local (freshId :) (go body)
+        Delay ann body -> do
+          freshId <- (+ 1) <$> lift get
+          lift $ put freshId
+          Delay (path, ann) <$> local (freshId :) (go body)
+        Case ann scrut branches ->
+          Case (path, ann)
+            <$> go scrut
+            <*> ( for branches $ \br -> do
+                    freshId <- (+ 1) <$> lift get
+                    lift $ put freshId
+                    local (freshId :) (go br)
+                )
+
+-- | The third pass. See Note [CSE].
+countOccs ::
+  forall name uni fun ann.
+  (Hashable (Term name uni fun ()), Hashable fun) =>
+  HashMap fun Int ->
+  Term name uni fun (Path, ann) ->
+  -- | Here, the value of the inner map not only contains the count, but also contains
+  -- the annotated term, corresponding to the term that is the key of the outer map.
+  -- The annotated terms need to be recorded since they will be used for substitution.
+  HashMap (Term name uni fun ()) (HashMap Path (Term name uni fun (Path, ann), Int))
+countOccs arityInfo = foldrOf termSubtermsDeep addToMap Map.empty
+  where
+    addToMap ::
+      Term name uni fun (Path, ann) ->
+      HashMap (Term name uni fun ()) (HashMap Path (Term name uni fun (Path, ann), Int)) ->
+      HashMap (Term name uni fun ()) (HashMap Path (Term name uni fun (Path, ann), Int))
+    addToMap t0
+      -- We don't consider work-free terms for CSE, because doing so may or may not
+      -- have a size benefit, but certainly doesn't have any cost benefit (the cost
+      -- will in fact be slightly higher due to the additional application).
+      --
+      -- `isWorkFree` currently doesn't check whether a builtin application is saturated,
+      -- or whether an term is the (possibly repeated) forcing of a builtin (which should
+      -- be workfree), so we check it separately.
+      | isWorkFree t0 || not (isBuiltinSaturated t0) || isForcingBuiltin t0 = id
```

Comment:
^ this comment is an important one. Perhaps you can even remove the entire first pass?

Reply:
I initially did this and it made things worse, I think because the on the one hand, we avoid evaluating `force someBuiltin` multiple times, but on the other hand, we evaluate `LamAbs` and `Apply` more times, which is more expensive!

Reply:
> we evaluate `LamAbs` and `Apply` more times, which is more expensive!

How come? Were you eta-expanding builtin applications? Or is it just the cost of creating a binding? Sounds strange to me.

Reply:
> How come? Were you eta-expanding builtin applications? Or is it just the cost of creating a binding? Sounds strange to me.

No, it's because when we CSE something, it creates an additional LamAbs and Apply.